File: src/main/java/com/funstat/ohlc/Sequenta.java
package com.funstat.ohlc;

import com.funstat.domain.Ohlc;
import com.iaa.finam.MdDao;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;


enum SignalType{
    SetupCount,SetupReach,SetupUnreach,Cdn,Recycling,Flip,Signal,Cancel,Expired,Deffered,Completed;
}

class Signal {
    SignalType type;
    Sequenta.Setup reference;
    Sequenta.Setup successor;

    public Signal(SignalType type, Sequenta.Setup reference) {
        this.type = type;
        this.reference = reference;
    }

    public Signal(SignalType type, Sequenta.Setup reference, Sequenta.Setup successor) {
        this.type = type;
        this.reference = reference;
        this.successor = successor;
    }

    @Override
    public String toString() {
        return "Signal{" +
                "type=" + type +
                '}';
    }
}



public class Sequenta {

    int[] counts = new int[]{13,21};

    List<Ohlc> data = new ArrayList<>();

    List<Setup> pendingSetups = new ArrayList<>();

    Ohlc past(int bars){
        return data.get(data.size() - bars);
    }

    class Setup{
        int start;
        int end;
        boolean up;
        List<Integer> countDowns = new ArrayList<>();
        int pendingSignal = 0;
        Setup cancelledRef;
        Setup recycleRef;
        double range = 0;

        void updateEnd(int idx){
            end = idx;
            range = calcRange();
        }

        double calcRange(){
            double min = Double.MAX_VALUE;
            double max = Double.MIN_VALUE;
            for(int i = start; i <= end; i++){
                min = Math.min(data.get(i).low, min);
                max = Math.max(data.get(i).high, max);
            }
            return max - min;

        }

        int length(){
            return end - start;
        }

        public Setup(int start, int end, boolean up) {
            this.start = start;
            this.end = end;
            this.up = up;
        }

        double recycleRatio(){
            return this.range/recycleRef.range;
        }

        boolean reached(){
            return (end - start) >= 9;
        }

        boolean isCancelled(){
            return cancelledRef != null;
        }

        boolean isCompleted(){
            return pendingSignal == counts.length;
        }

        boolean isCurrent(){
            return this == pendingSetups.get(pendingSetups.size() - 1);
        }

        List<Signal> checkCountDown(){
            int idx = data.size() - 1;
            List<Signal> ret = new ArrayList<>();
            if(cdn(idx)){
                countDowns.add(idx);
                ret.add(new Signal(SignalType.Cdn, this));
            }
            if(countDowns.size() >= counts[pendingSignal]){
                if(up && data.get(idx).high > data.get(countDowns.get(8)).close
                        || !up && data.get(idx).low < data.get(countDowns.get(8)).close
                        ){
                    pendingSignal++;
                    ret.add(new Signal(SignalType.Signal, this));
                }else {
                    ret.add(new Signal(SignalType.Deffered, this));
                }
            }
            if(pendingSignal == counts.length){
                ret.add(new Signal(SignalType.Completed, this));
            }
            return ret;
        }


        private boolean cdn(int idx) {
            return up && getClose(idx) > data.get(idx - 2).high
                    || !up && getClose(idx) < data.get(idx - 2).low;
        }

        private double getClose(int idx) {
            return data.get(idx).close;
        }
    }

    Ohlc last(int idx){
        return data.get(data.size() - idx);
    }

    boolean getCurrentTrend(){
        Ohlc ohlc = data.get(data.size() - 1);
        return ohlc.close > past(4).close;
    }

    Setup currentSetup;

    public List<Signal> onOhlc(Ohlc ohlc){
        data.add(ohlc);
        if(data.size() < 4){
            return Collections.emptyList();
        }
        if(data.size() == 4){
            currentSetup = new Setup(data.size() - 1, data.size() - 1, getCurrentTrend());
            return Collections.emptyList();
        }
        List<Signal> ret = new ArrayList();
        ret.addAll(runCurrent());

        this.pendingSetups.forEach(ps->{
            ret.addAll(ps.checkCountDown());
        });
        this.pendingSetups = this.pendingSetups.stream().filter(ps->{
            return !ps.isCompleted();
        }).collect(Collectors.toList());
        return ret;
    }


    private List<Signal> runCurrent() {
        List<Signal> ret = new ArrayList<>();
        int idx = data.size() - 1;
        currentSetup.updateEnd(idx);
        currentSetup.end = idx;
        if( getCurrentTrend() != currentSetup.up){
            if(currentSetup.length() < 9){
                ret.add(new Signal(SignalType.SetupUnreach, currentSetup));
            }else {
                ret.add(new Signal(SignalType.Flip, currentSetup));
            }
            currentSetup = new Setup(idx,idx, getCurrentTrend());
            return ret;
        }else {
            ret.add(new Signal(SignalType.SetupCount, currentSetup));
        }
        if(currentSetup.length() == 9){
            pendingSetups.add(currentSetup);
            ret.add(new Signal(SignalType.SetupReach, currentSetup));
        }

        for(int i = 0; i < pendingSetups.size() - 1; i++){
            Setup tchk = pendingSetups.get(i);
            if(tchk == currentSetup){
                break;
            }
            if(currentSetup.up == tchk.up && currentSetup.range > tchk.range){
                if(tchk.recycleRef == null || tchk.recycleRef.range < currentSetup.range){
                    tchk.recycleRef = currentSetup;
                    ret.add(new Signal(SignalType.Recycling, tchk, currentSetup));
                }
            }

            if(currentSetup.reached() && currentSetup.up != tchk.up){
                if(tchk.cancelledRef== null || tchk.cancelledRef.range < currentSetup.range){
                    tchk.cancelledRef = currentSetup;
                    ret.add(new Signal(SignalType.Cancel, tchk, currentSetup));
                }
            }
        }
        return ret;
    }



    public static void testSequenta(){
        Sequenta sequenta = new Sequenta();
        List<Ohlc> testHistory = new ArrayList<>();

        sequenta.onOhlc(new Ohlc(LocalDateTime.now(), 1, 2, 0, 1 ));
        sequenta.onOhlc(new Ohlc(LocalDateTime.now(), 1, 2, 0, 1 ));
        sequenta.onOhlc(new Ohlc(LocalDateTime.now(), 1, 2, 0, 1 ));
        sequenta.onOhlc(new Ohlc(LocalDateTime.now(), 1, 2, 0, 1 ));

        for(int i = 1; i < 10; i++){
            sequenta.onOhlc(new Ohlc(LocalDateTime.now(), i, i - 1, i, i + 1));
        }

        for (int i = 0; i < 12; i++){
            Ohlc last = sequenta.last(2);
            sequenta.onOhlc(last);
            sequenta.onOhlc(new Ohlc(last.dateTime, last.close,last.high + 0.2,last.low - 0.2,last.high + 0.1));
        }

        Ohlc last = sequenta.last(2);
        sequenta.onOhlc(last);
        List<Signal> signals = sequenta.onOhlc(new Ohlc(last.dateTime, last.close,last.high + 0.2,last.low - 0.2,last.high + 0.1));
        System.out.println(signals);
    }

    public static void main(String[] args) {
        MdDao mdDao = new MdDao();
        List<Ohlc> tatn = mdDao.queryAll("TATN");
        Sequenta sequenta = new Sequenta();
        tatn.forEach(oh->{
            System.out.println(sequenta.onOhlc(oh));
        });
    }

}

File: back/src/main/java/firelib/indicators/sequenta/Sequenta.kt
package firelib.indicators.sequenta

import firelib.core.domain.Ohlc
import firelib.core.domain.range
import java.time.Instant


class Sequenta(val counts : Array<Int> = arrayOf(13, 21)) {

    var data: MutableList<Ohlc> = ArrayList()

    var pendingSetups: MutableList<Setup> = ArrayList()

    val currentTrend: Boolean
        get() {
            return past(1).close > past(5).close
        }

    var currentSetup: Setup = Setup(0, 0, false)

    fun past(bars: Int): Ohlc {
        return data[data.size - bars]
    }

    inner class Setup(var start: Int, var end: Int, var up: Boolean) {
        var countDowns: MutableList<Int> = ArrayList()
        var pendingSignal = 0
        var cancelledRef: Setup? = null
        var recycleRef: Setup? = null

        val isUp get() =  up
        val isDown get() = !up

        var closesBeyondTdst = 0

        var min: Double = 0.toDouble()
        var max: Double = 0.toDouble()

        val tdst: Double
            get() = if (up) min else max

        val isCancelled: Boolean
            get() = cancelledRef != null

        val isCompleted: Boolean
            get() = pendingSignal == counts.size

        val isExpired: Boolean
            get() = data.size - start > 150


        val completedSignal: Int
            get() = if (pendingSignal == 0) {
                -1
            } else counts[pendingSignal - 1]


        fun data() : List<Ohlc>{
            return data
        }

        fun updateEnd(idx: Int) {
            end = idx
            calcRange()
        }

        fun getStart(): Instant {
            return data[start].endTime
        }

        fun getEnd(): Instant {
            return data[end].endTime
        }

        fun setupSize(): Int {
            return end - start
        }


        fun calcRange() {
            this.min = java.lang.Double.MAX_VALUE
            this.max = java.lang.Double.MIN_VALUE
            for (i in start..end) {
                min = Math.min(data[i].low, min)
                max = Math.max(data[i].high, max)
            }
        }

        fun range(): Double {
            return max - min
        }

        fun length(): Int {
            return end - start
        }

        fun recycleRatio(): Double? {
            return if (recycleRef == null) {
                null
            } else recycleRef!!.range() / this.range()
        }

        fun reached(): Boolean {
            return end - start >= SETUP_LENGTH - 1
        }

        fun invalidated(): Boolean {
            return closesBeyondTdst > 5
        }

        fun sequenta() : Sequenta{
            return this@Sequenta
        }

        fun checkCountDown(): List<Signal> {
            val idx = data.size - 1
            val ret = ArrayList<Signal>()

            checkClosesBeyondTdst(idx)

            if (isCntdn(idx)) {
                countDowns.add(idx)
                ret.add(Signal(SequentaSignalType.Cdn, this))
                if (countDowns.size >= counts[pendingSignal]) {
                    if (up && data[idx].high > data[countDowns[7]].close || !up && data[idx].low < data[countDowns[7]].close) {
                        pendingSignal++
                        ret.add(Signal(SequentaSignalType.Signal, this))
                    } else {
                        ret.add(Signal(SequentaSignalType.Deffered, this))
                    }
                }
            }
            if (isCompleted) {
                ret.add(Signal(SequentaSignalType.Completed, this))
            }
            return ret
        }

        private fun checkClosesBeyondTdst(idx: Int) {
            if (up) {
                if (data[idx].close < tdst) {
                    closesBeyondTdst++
                }
            } else {
                if (data[idx].close > tdst) {
                    closesBeyondTdst++
                }
            }
        }


        private fun isCntdn(idx: Int): Boolean {
            return up && getClose(idx) > data[idx - 2].high || !up && getClose(idx) < data[idx - 2].low
        }

        private fun getClose(idx: Int): Double {
            return data[idx].close
        }
    }

    fun last(idx: Int): Ohlc {
        return data[data.size - idx]
    }

    fun onOhlc(ohlc: Ohlc): List<Signal> {
        data.add(ohlc)
        if (data.size < 5) {
            return emptyList()
        }
        if (data.size == 5) {
            currentSetup = Setup(data.size - 1, data.size - 1, currentTrend)
            return emptyList()
        }
        val ret = runCurrent().toMutableList()

        this.pendingSetups.forEach { ps -> ret.addAll(ps.checkCountDown()) }
        this.pendingSetups =
            this.pendingSetups.filter { !it.isCompleted && !it.isExpired && !it.invalidated() }.toMutableList()
        return ret
    }


    private fun runCurrent(): List<Signal> {
        val ret = ArrayList<Signal>()
        val idx = data.size - 1
        currentSetup.updateEnd(idx)
        if (currentTrend != currentSetup.up) {
            if (!currentSetup.reached()) {
                ret.add(Signal(SequentaSignalType.SetupUnreach, currentSetup))
            } else {
                ret.add(Signal(SequentaSignalType.Flip, currentSetup))
            }
            currentSetup = Setup(idx, idx, currentTrend)
            return ret
        } else {
            ret.add(Signal(SequentaSignalType.SetupCount, currentSetup))
        }
        //check completed because sometimes setup not flipped untill 21
        if (currentSetup.reached() && !pendingSetups.contains(currentSetup) && !currentSetup.isCompleted) {
            pendingSetups.add(currentSetup)
            ret.add(Signal(SequentaSignalType.SetupReach, currentSetup))
        }

        for (i in 0 until pendingSetups.size - 1) {
            val tchk = pendingSetups[i]
            if (tchk === currentSetup) {
                break
            }
            if (currentSetup.up == tchk.up && currentSetup.range() > tchk.range()) {
                if (tchk.recycleRef == null || tchk.recycleRef!!.range() < currentSetup.range()) {
                    tchk.recycleRef = currentSetup
                    ret.add(Signal(SequentaSignalType.Recycling, tchk, currentSetup))
                }
            }

            if (currentSetup.reached() && currentSetup.up != tchk.up) {
                if (tchk.cancelledRef == null || tchk.cancelledRef!!.range() < currentSetup.range()) {
                    tchk.cancelledRef = currentSetup
                    ret.add(Signal(SequentaSignalType.Cancel, tchk, currentSetup))
                }
            }
        }
        return ret
    }

    companion object {
        val SETUP_LENGTH = 9
    }
}


fun List<Ohlc>.calcStop(up : Boolean, start : Int, end : Int): Double {
    var curLevel: Double
    if (up) {
        curLevel = java.lang.Double.MIN_VALUE
        for (i in start until end) {
            val ohh = this[i]
            curLevel = Math.max(curLevel, ohh.high + ohh.range())
        }
    } else {
        curLevel = java.lang.Double.MAX_VALUE
        for (i in start until end) {
            val ohh = this[i]
            curLevel = Math.min(curLevel, ohh.low - ohh.range())
        }
    }
    return curLevel
}